src/utils/signals.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {map} from './object';
+
+
+/**
+ * This object tracts signals and allows blocking until a signal has been
+ * received.
+ */
+export class Signals {
```

Comment:
This is the cost of being able to track past events that maybe requested in the future. Can't see another option, tbh. But would be happy to discuss.